text/0001-dynamic-fees-v1.md

```diff
@@ -0,0 +1,93 @@
+
+If two nodes with different fees attempt to communicate, the result will be an `AeadError` during the handshake phase.
+
+The currently-configured fee should be made available to consensus clients directly by extending the `BlockchainAPI.GetLastBlockInfo` response. This gRPC call is already polled by thick clients to orient themselves with relation to the state of the network.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `consensus_service` application will be extended to add a new `--minimum-fee` command-line argument and `MC_MINIMUM_FEE` environment variable. These values will be merged (environment-variable, then command-line option). This value will contain a string indicating the quantity and SI-scale of the minimum fee. Fees with out units will be considered to be denominated in picoMOB. Therefore, all these values are equivilent fees:
+
+- 10mMOB
+- 10000uMOB
+- 10_000_000_000
+
+SI units will be parsed in a case-sensitive manner following the [NIST SI Prefixes](https://www.nist.gov/pml/weights-and-measures/metric-si-prefixes). Most importantly, GMOB will be correctly pronounced as "JIG-uh MOEB" (with a soft "g").
+
+Fractional fees (e.g. "0.01MOB") will be rejected, to avoid the potential for floating-point ambiguity in implementation. Additionally, the consensus service will reject any fees larger than 1 MOB or smaller than 0.00000001 MOB (0.005 USD @ 50 kUSD), unless the operator also includes an `--allow-any-fees` command-line option. This is intended to mitigate some of the risk of SI unit confusion (e.g. "MMOB", indicating millions of MOB, vs. "mMOB", indicating thousandths of a MOB.
+
+The given minimum fee value, if it is provided, will be injected into the enclave at startup and stored for future use when validating inbound transactions. As is the case today, fees will be validated initially, on transaction submission.
+
+This fee value will be injected into the attested key exchange by the enclave, by constructing a new UTF-8 string of the format: `"%s-%u"` using the existing `ResponderId`, and the u64 picoMOB-denominated representation of the new fee. This value is injected into the hashing scheme built onto the key exchange protocols in use.
+
+Because of this, two nodes with different fees will produce different encryption keys, which will lead to a failed decyption during the key exchange setup (specifically, the initiator will receive an `AeadError` when attempting to decrypt the initial message received from the responder).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The primary drawback to this approach is that it requires a restart of a quorum of the network in order to adjust to a new fee. In the intended common-case (low fees generally, higher fees in response to an active denial-of-service), this will entail restarting the network twice, and when a node is restarted, any pending transactions which did not make it into the current SCP slot will be discarded.
+
```

Comment:
If it's not possible to be more descriptive in the `AeadError`, which it seems like it might not be, then I would add the challenge of debugging the attestation failure as a drawback.

Reply:
Fair point.